# 4.7 Deriving New Variables

### This script contains the following points:

#### 01. Importing Libraries and Loading Files
#### 02. User-Defined Functions
#### 03. If-Statements with the loc() Function
#### 04. If-Statement with For-Loops
#### 05. Task Exercises

### 01. Importing Libraries and Loading Files

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

# Creating path variable to load files
path = r'C:\Users\widne\Documents\CareerFoundry Exercises\Data_Immersion\Achievement 4\03-2024 Instacart Basket Analysis'

# Loading orders_products_merged pickle as df_ords_prods_merge dataframe
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

### 02. User-Defined Functions

In [8]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  first_order  add_to_cart_order  reordered _merge  
0                     3.0        False                  5          0   both  
1                    20.0        False                  1          1   both  
2                     6.0        False                 20          0   both  
3                     NaN         True                 10          0   both  
4                     9.0        False                 11          1   both

In [2]:
# Creating subset of df_ords_prods_merge
df = df_ords_prods_merge[:1000000]

# Price label function
def price_label(row):
    if row['prices'] <= 5:
        return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High range product'
    else: return 'Not enough data'

# Running price_label function
df['price_range'] = df.apply(price_label, axis = 1)

C:\Users\widne\AppData\Local\Temp\ipykernel_20788\3702974691.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis = 1)


In [3]:
df.shape

(1000000, 16)

In [4]:
# Checking values of new column
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [5]:
df['prices'].max()

14.8

### 03. If-Statements with the loc() Function

In [6]:
# Using loc() function in if-statements
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\widne\AppData\Local\Temp\ipykernel_20788\1617844576.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [7]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [8]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [9]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [13]:
# Using loc() function on df_ords_prods_merge
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [16]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [17]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [18]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

### 04. If-Statement with For-Loops

In [20]:
# Finding busiest day
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [21]:
# Creating for loop
result = []

for value in df_ords_prods_merge['orders_day_of_week']:
    if value == 0:
        result.append('Busiest day')
    elif value == 4:
        result.append('Least busy')
    else:
        result.append('Regularly busy')

In [22]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Bus

In [23]:
# Creating new result column in df_ords_prods_merge
df_ords_prods_merge['busiest_day'] = result

In [24]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

### 05. Task Exercises

In [28]:
# Creating busiest_days colunmn that labels the two busiest days as busiest and the two least busy as least busy
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

The two busiest days are Saturday (0) and Sunday (1). The two least busy days are Tuesday (3) and Wednesday (4)

In [3]:
result = []

# Creating for loop to check for busiest days
for value in df_ords_prods_merge['order_day_of_week']:
    if (value == 0) or (value == 1):
        result.append('Busiest days')
    elif (value == 3) or (value == 4):
        result.append('Least busy')
    else:
        result.append('Regularly busy')

# Creating new column named busiest_days
df_ords_prods_merge['busiest_days'] = result

In [4]:
# Checking values of busiest_days column
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12916111
Busiest days      11864412
Least busy         7624336
Name: count, dtype: int64

Busiest Days: Saturday + Sunday (6204182 + 5660230) = 11864412 which is the Busiest days count
Least Busy Days: Tuesday + Wednesday (3840534 + 3783802) = 7624336 which is the Least busy days count

In [5]:
# Identifying the busiest times of day
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

For identifying the busiest times of day there will be 3 labels:
The top 8 order_hour_of_day frequencies will be labeled as Most orders (9, 10, 11, 12, 13, 14, 15, 16).
The bottom 8 order_hour_of_day frequencies will be labeled as Fewest orders (0, 1, 2, 3, 4, 5, 6, 23).
Everything else will be labeled as Average orders

In [6]:
result = []

# Creating for loop to get busiest hours of day
for value in df_ords_prods_merge['order_hour_of_day']:
    if value in [9, 10, 11, 12, 13, 14, 15, 16]:
        result.append('Most orders')
    elif value in [0, 1, 2, 3, 4, 5, 6, 23]:
        result.append('Fewest orders')
    else: 
        result.append('Average orders')

result

['Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most ord

In [7]:
# Creating new column named busiest_period_of_day
df_ords_prods_merge['busiest_period_of_day'] = result

In [8]:
# Checking values of new column
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

In [9]:
# Exporting df_ords_prods_merge as pickle
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))